<a href="https://colab.research.google.com/github/SI-GRush34/beijing-air-quality-project/blob/main/Beijing-Air-Quality-Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Beijing Air Quality Analysis
## Task 1 – Data Selection and Handling

High concentrations of air pollutants like PM2.5, PM10, SO2, NO2, CO, and O3 can cause serious respiratory and cardiovascular diseases. Beijing, the capital city of :contentReference[oaicite:0]{index=0}, has had serious air pollution issues over the past 20 years due to rapid urbanisation, industrialisation, and rising vehicle emissions. The goal of this task is to select and integrate air quality datasets collected from monitoring stations in :contentReference[oaicite:1]{index=1}. Severe respiratory and cardiovascular conditions can result from high levels of air pollutants such PM2.5, PM10, SO2, NO2, CO, and O3. Over the past 20 years, increased urbanisation, industrialisation, and rising car emissions have caused serious air pollution issues in Beijing, the capital city of :contentReference[oaicite:0]{index=0}.

The selection and integration of air quality datasets gathered from monitoring stations in :contentReference[oaicite:1]{index=1} is the main goal of this assignment. The goal is to import, preprocess, and combine several datasets into a single, cohesive dataset that may be used for predictive modelling and additional exploratory data analysis.

## Importing Required Libraries

To facilitate data preprocessing, analysis, and visualisation chores, the following Python libraries were imported: Matplotlib and Seaborn are used for visualisation, while Pandas and NumPy are used for data manipulation and numerical calculations. Matplotlib and Seaborn are used for visualisation, while Pandas and NumPy are utilised for numerical computations and data manipulation.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from glob import glob

## Monitoring Station Selection From the Beijing air quality monitoring network, two urban and two suburban monitoring stations were chosen for this analysis

From the Beijing air quality monitoring network, two urban and two suburban monitoring sites were chosen for this investigation:
Urban Stations: Dongsi and Guanyuan,
Suburban Stations: Changping and Huairou

Dongsi and Guanyuan were chosen as urban stations because they are situated in densely populated central regions of :contentReference[oaicite:2]{index=2}, where traffic congestion, commercial activity, and industrial emissions are relatively high. As a result, these locations are representative of inner-city pollution conditions. Changping and Huairou were chosen as suburban stations because they are located farther from the urban center and are characterised by lower population density and reduced industrial activity. These stations offer a useful comparison between urban and suburban air quality conditions. The categorisation was supported by earlier research, including studies by Xu & Zhang and Yao et al. As a result, these places are typical of pollution levels in inner cities.

Due to their distance from the city center, lower population density, and less industrial activity, Changping and Huairou were chosen as suburban stations. These stations offer a helpful comparison of the quality of the air in cities and suburbs.

Previous research, such as studies by Xu & Zhang and Yao et al., which categorise monitoring locations based on urban and suburban features, was used to support the station classification.

## Importing Selected Air Quality Datasets

The chosen datasets, which include hourly meteorological and air quality measurements gathered between March 2013 and February 2017, were uploaded into the Google Colab environment and imported using the Pandas library.

The Pandas library was used to import the chosen datasets after they were uploaded into the Google Colab environment. Hourly meteorological and air quality data from March 2013 to February 2017 are included in each file.

In [2]:
import os
import pandas as pd

# Get file names
files = os.listdir('/content')

# Create DataFrame
files_df = pd.DataFrame(files, columns=['Available Dataset Files Below'])

# Add numbering
files_df.index = range(1, len(files_df) + 1)

# Display
files_df.style.set_properties(**{
    'text-align': 'left',
    'font-size': '12pt'
}).set_caption("** List of Available Files in the Working Directory **")

,Available Dataset Files Below
1,.config
2,mnist_train_small.csv
3,california_housing_test.csv
4,README.md
5,PRSA_Data_Dongsi_20130301-20170228.csv
6,sample_data
7,mnist_test.csv
8,PRSA_Data_Changping_20130301-20170228.csv
9,PRSA_Data_Guanyuan_20130301-20170228.csv
10,anscombe.json


In [3]:
import pandas as pd

def load_dataset(file_path, station_name):
    df = pd.read_csv(file_path)
    df['station'] = station_name
    return df

dongsi = pd.read_csv('/content/PRSA_Data_Dongsi_20130301-20170228.csv')
guanyuan = pd.read_csv('/content/PRSA_Data_Guanyuan_20130301-20170228.csv')
changping = pd.read_csv('/content/PRSA_Data_Changping_20130301-20170228.csv')
huairou = pd.read_csv('/content/PRSA_Data_Huairou_20130301-20170228.csv')

print("All the datasets loaded successfully!!!")

All the datasets loaded successfully!!!


## Dataset Preview

The first 5 rows of the Dongsi monitoring station dataset are shown in the output that follows. This preview aids in confirming that the dataset has been successfully imported into the development environment.

In [4]:
print("Dongsi Dataset")
display(dongsi.head())

print("Guanyuan Dataset")
display(guanyuan.head())

print("Changping Dataset")
display(changping.head())

print("Huairou Dataset")
display(huairou.head())

print(dongsi.shape)

Dongsi Dataset


,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,9.0,9.0,3.0,17.0,300.0,89.0,-0.5,1024.5,-21.4,0.0,NNW,5.7,Dongsi
1,2,2013,3,1,1,4.0,4.0,3.0,16.0,300.0,88.0,-0.7,1025.1,-22.1,0.0,NW,3.9,Dongsi
2,3,2013,3,1,2,7.0,7.0,NaN,17.0,300.0,60.0,-1.2,1025.3,-24.6,0.0,NNW,5.3,Dongsi
3,4,2013,3,1,3,3.0,3.0,5.0,18.0,NaN,NaN,-1.4,1026.2,-25.5,0.0,N,4.9,Dongsi
4,5,2013,3,1,4,3.0,3.0,7.0,NaN,200.0,84.0,-1.9,1027.1,-24.5,0.0,NNW,3.2,Dongsi


Guanyuan Dataset


,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,4.0,4.0,14.0,20.0,300.0,69.0,-0.7,1023.0,-18.8,0.0,NNW,4.4,Guanyuan
1,2,2013,3,1,1,4.0,4.0,13.0,17.0,300.0,72.0,-1.1,1023.2,-18.2,0.0,N,4.7,Guanyuan
2,3,2013,3,1,2,3.0,3.0,10.0,19.0,300.0,69.0,-1.1,1023.5,-18.2,0.0,NNW,5.6,Guanyuan
3,4,2013,3,1,3,3.0,6.0,7.0,24.0,400.0,62.0,-1.4,1024.5,-19.4,0.0,NW,3.1,Guanyuan
4,5,2013,3,1,4,3.0,6.0,5.0,14.0,400.0,71.0,-2.0,1025.2,-19.5,0.0,N,2.0,Guanyuan


Changping Dataset


,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,3.0,6.0,13.0,7.0,300.0,85.0,-2.3,1020.8,-19.7,0.0,E,0.5,Changping
1,2,2013,3,1,1,3.0,3.0,6.0,6.0,300.0,85.0,-2.5,1021.3,-19.0,0.0,ENE,0.7,Changping
2,3,2013,3,1,2,3.0,3.0,22.0,13.0,400.0,74.0,-3.0,1021.3,-19.9,0.0,ENE,0.2,Changping
3,4,2013,3,1,3,3.0,6.0,12.0,8.0,300.0,81.0,-3.6,1021.8,-19.1,0.0,NNE,1.0,Changping
4,5,2013,3,1,4,3.0,3.0,14.0,8.0,300.0,81.0,-3.5,1022.3,-19.4,0.0,N,2.1,Changping


Huairou Dataset


,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,7.0,7.0,3.0,2.0,100.0,91.0,-2.3,1020.3,-20.7,0.0,WNW,3.1,Huairou
1,2,2013,3,1,1,4.0,4.0,3.0,NaN,100.0,92.0,-2.7,1020.8,-20.5,0.0,NNW,1.5,Huairou
2,3,2013,3,1,2,4.0,4.0,NaN,NaN,100.0,91.0,-3.2,1020.6,-21.4,0.0,NW,1.8,Huairou
3,4,2013,3,1,3,3.0,3.0,3.0,2.0,NaN,NaN,-3.3,1021.3,-23.7,0.0,NNW,2.4,Huairou
4,5,2013,3,1,4,3.0,3.0,7.0,NaN,300.0,86.0,-4.1,1022.1,-22.7,0.0,NNW,2.2,Huairou


(35064, 18)


## Comprehending Dataset Structure

Dataset Information:

Prior to carrying out preprocessing and merging procedures, it is crucial to comprehend the dataset structure, which was analysed to determine the number of records, column names, data kinds, and missing values.

In [5]:
def styled_summary(df, name):
    summary = pd.DataFrame({
        'Column Name': df.columns,
        'Data Type': df.dtypes.values,
        'Missing Values': df.isnull().sum().values,
        'Non-Null Count': df.notnull().sum().values
    })

    display(
        summary.style
        .set_caption(f"{name} Dataset Information Summary")
        .background_gradient(cmap='Blues')
        .set_properties(**{
            'text-align': 'center',
            'font-size': '11pt'
        })
    )

styled_summary(dongsi, "Dongsi")
styled_summary(guanyuan, "Guanyuan")
styled_summary(changping, "Changping")
styled_summary(huairou, "Huairou")

,Column Name,Data Type,Missing Values,Non-Null Count
0,No,int64,0,35064
1,year,int64,0,35064
2,month,int64,0,35064
3,day,int64,0,35064
4,hour,int64,0,35064
5,PM2.5,float64,750,34314
6,PM10,float64,553,34511
7,SO2,float64,663,34401
8,NO2,float64,1601,33463
9,CO,float64,3197,31867


,Column Name,Data Type,Missing Values,Non-Null Count
0,No,int64,0,35064
1,year,int64,0,35064
2,month,int64,0,35064
3,day,int64,0,35064
4,hour,int64,0,35064
5,PM2.5,float64,616,34448
6,PM10,float64,429,34635
7,SO2,float64,474,34590
8,NO2,float64,659,34405
9,CO,float64,1753,33311


,Column Name,Data Type,Missing Values,Non-Null Count
0,No,int64,0,35064
1,year,int64,0,35064
2,month,int64,0,35064
3,day,int64,0,35064
4,hour,int64,0,35064
5,PM2.5,float64,774,34290
6,PM10,float64,582,34482
7,SO2,float64,628,34436
8,NO2,float64,667,34397
9,CO,float64,1521,33543


,Column Name,Data Type,Missing Values,Non-Null Count
0,No,int64,0,35064
1,year,int64,0,35064
2,month,int64,0,35064
3,day,int64,0,35064
4,hour,int64,0,35064
5,PM2.5,float64,953,34111
6,PM10,float64,777,34287
7,SO2,float64,980,34084
8,NO2,float64,1639,33425
9,CO,float64,1422,33642


## Finding Missing Values

In order to assess the quality of the data and ascertain whether preprocessing methods would be necessary prior to analysis and predictive modelling, missing values were looked at.

In [6]:
import pandas as pd

# Function to display missing values summary
def missing_values_summary(df, name):

    missing_df = pd.DataFrame({
        'Column Name': df.columns,
        'Missing Values': df.isnull().sum().values
    })

    print(f"\n{name} Dataset - Missing Values Summary")

    display(
        missing_df.style
        .background_gradient(cmap='Reds')
        .set_caption(f"{name} Missing Values Report")
        .set_properties(**{
            'text-align': 'center',
            'font-size': '11pt'
        })
    )

# Display summaries
missing_values_summary(dongsi, "Dongsi")
missing_values_summary(guanyuan, "Guanyuan")
missing_values_summary(changping, "Changping")
missing_values_summary(huairou, "Huairou")


Dongsi Dataset - Missing Values Summary


,Column Name,Missing Values
0,No,0
1,year,0
2,month,0
3,day,0
4,hour,0
5,PM2.5,750
6,PM10,553
7,SO2,663
8,NO2,1601
9,CO,3197



Guanyuan Dataset - Missing Values Summary


,Column Name,Missing Values
0,No,0
1,year,0
2,month,0
3,day,0
4,hour,0
5,PM2.5,616
6,PM10,429
7,SO2,474
8,NO2,659
9,CO,1753



Changping Dataset - Missing Values Summary


,Column Name,Missing Values
0,No,0
1,year,0
2,month,0
3,day,0
4,hour,0
5,PM2.5,774
6,PM10,582
7,SO2,628
8,NO2,667
9,CO,1521



Huairou Dataset - Missing Values Summary


,Column Name,Missing Values
0,No,0
1,year,0
2,month,0
3,day,0
4,hour,0
5,PM2.5,953
6,PM10,777
7,SO2,980
8,NO2,1639
9,CO,1422


To identify the matching air quality monitoring station, a new column called station was introduced to each dataset. When combining several datasets, this phase is crucial since it allows for comparison analysis between various monitoring locations while maintaining the source information of each observation.

In [7]:
dongsi['station'] = 'Dongsi'
guanyuan['station'] = 'Guanyuan'
changping['station'] = 'Changping'
huairou['station'] = 'Huairou'

In order to preserve station-specific observations while guaranteeing that only true duplicate rows are eliminated from the dataset, the datasets from various monitoring stations were first combined into a single DataFrame using pd.concat(). Duplicate records were then found using the duplicated() function and eliminated using drop_duplicates().

In [8]:
# Combine all datasets
combined_df = pd.concat([dongsi, guanyuan, changping, huairou], ignore_index=True)

# Duplicate rows
print("Number of duplicate rows:", combined_df.duplicated().sum())

# Remove duplicate rows
combined_df.drop_duplicates(inplace=True)

# Verify duplicates after removal
print("Duplicate rows after removal:", combined_df.duplicated().sum())

Number of duplicate rows: 0
Duplicate rows after removal: 0


## Handling Timestamps



The year, month, day, and hour columns in the original dataset each contain temporal information separately. To enhance time-series analysis and streamline merging procedures, these columns were merged into a single datetime column using Pandas datetime capabilities.

In [9]:
import pandas as pd

# Function to create datetime column
def create_datetime(df):
    df['datetime'] = pd.to_datetime(
        df[['year', 'month', 'day', 'hour']]
    )
    return df

# Apply function
dongsi = create_datetime(dongsi)
guanyuan = create_datetime(guanyuan)
changping = create_datetime(changping)
huairou = create_datetime(huairou)

# Add station labels
dongsi['station'] = 'Dongsi'
guanyuan['station'] = 'Guanyuan'
changping['station'] = 'Changping'
huairou['station'] = 'Huairou'

# Merge datasets
combined_df = pd.concat(
    [dongsi, guanyuan, changping, huairou],
    ignore_index=True
)

timestamp_preview = combined_df[['datetime', 'station']].head(10)
display(
    timestamp_preview.style
    .set_caption("📅 Air Quality Timestamp Handling with Station Info")
    .background_gradient(cmap="plasma")
    .set_properties(**{
        'text-align': 'center',
        'font-size': '12pt'
    })
    .highlight_max(axis=0, color='skyblue')
)

,datetime,station
0,2013-03-01 00:00:00,Dongsi
1,2013-03-01 01:00:00,Dongsi
2,2013-03-01 02:00:00,Dongsi
3,2013-03-01 03:00:00,Dongsi
4,2013-03-01 04:00:00,Dongsi
5,2013-03-01 05:00:00,Dongsi
6,2013-03-01 06:00:00,Dongsi
7,2013-03-01 07:00:00,Dongsi
8,2013-03-01 08:00:00,Dongsi
9,2013-03-01 09:00:00,Dongsi


In [10]:
def create_datetime(df):
    df['datetime'] = pd.to_datetime(
        df[['year', 'month', 'day', 'hour']]
    )
    return df

In [11]:
dongsi = create_datetime(dongsi)
guanyuan = create_datetime(guanyuan)
changping = create_datetime(changping)
huairou = create_datetime(huairou)

In [12]:
dongsi[['datetime']].head()

,datetime
0,2013-03-01 00:00:00
1,2013-03-01 01:00:00
2,2013-03-01 02:00:00
3,2013-03-01 03:00:00
4,2013-03-01 04:00:00


## Combining and Merging Certain Datasets

The Pandas concat() function was used to combine the four chosen datasets into a single, cohesive dataframe. Comparative study between urban and suburban monitoring stations is made possible by combining the datasets.

To ensure data persistence for future analysis, the merged dataset was saved as a CSV file using the to_csv() function. The structure of the dataset was then confirmed using the.shape attribute, which provides the total number of rows and columns after merging multiple station datasets. Lastly, the first five records were displayed using the.head() function to confirm successful integration and to provide an initial view of the combined air quality dataset. The results were presented in a formatted and visually enhanced table to improve readability and interpretability.

In [13]:
dongsi['station'] = 'Dongsi'
guanyuan['station'] = 'Guanyuan'
changping['station'] = 'Changping'
huairou['station'] = 'Huairou'

In [14]:
import pandas as pd

# Save merged dataset to CSV
combined_df.to_csv('combined_air_quality.csv', index=False)

# Print dataset shape
shape_df = pd.DataFrame({
    "Metric": ["Rows", "Columns"],
    "Value": combined_df.shape
}).T

print("📌 Merged Dataset Overview")

display(
    shape_df.style
    .set_caption("📊 Dataset Shape Summary")
    .background_gradient(cmap="Blues")
    .set_properties(**{
        "text-align": "center",
        "font-size": "12pt"
    })
)

# Show first 5 rows
display(
    combined_df.head().style
    .set_caption("🌫️ First 5 Records of Combined Air Quality Dataset")
    .background_gradient(cmap="Greens")
)

📌 Merged Dataset Overview


,0,1
Metric,Rows,Columns
Value,140256,19


,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station,datetime
0,1,2013,3,1,0,9.000000,9.000000,3.000000,17.000000,300.000000,89.000000,-0.500000,1024.500000,-21.400000,0.000000,NNW,5.700000,Dongsi,2013-03-01 00:00:00
1,2,2013,3,1,1,4.000000,4.000000,3.000000,16.000000,300.000000,88.000000,-0.700000,1025.100000,-22.100000,0.000000,NW,3.900000,Dongsi,2013-03-01 01:00:00
2,3,2013,3,1,2,7.000000,7.000000,nan,17.000000,300.000000,60.000000,-1.200000,1025.300000,-24.600000,0.000000,NNW,5.300000,Dongsi,2013-03-01 02:00:00
3,4,2013,3,1,3,3.000000,3.000000,5.000000,18.000000,nan,nan,-1.400000,1026.200000,-25.500000,0.000000,N,4.900000,Dongsi,2013-03-01 03:00:00
4,5,2013,3,1,4,3.000000,3.000000,7.000000,nan,200.000000,84.000000,-1.900000,1027.100000,-24.500000,0.000000,NNW,3.200000,Dongsi,2013-03-01 04:00:00


## Task 2 – Exploratory Data Analysis (EDA)

# 2.1 Data Understanding

The Beijing Air Quality dataset was used in this study to find patterns, identify missing values, comprehend pollutant distributions, and obtain insights into environmental conditions. Exploratory Data Analysis (EDA) is a crucial step in the data analysis process that is used to understand the structure, quality, and characteristics of the dataset before applying machine learning or statistical techniques. To find trends, find missing information, comprehend pollutant distributions, and learn more about environmental conditions, this study used EDA on the Beijing Air Quality dataset.

#Import Libraries

The necessary Python libraries for data analysis and visualisation throughout the project are imported in this part. These libraries include features for numerical calculations, data manipulation, and graphical display of the air quality dataset.

Importing these libraries at the start of the notebook helps maintain an organised workflow and guarantees that all necessary tools are available for subsequent preprocessing, analysis, and model development tasks.

- `pandas` is used for handling and analysing structured data in tabular format;
- `numpy` supports numerical and mathematical operations efficiently;
- `matplotlib.pyplot` is used for creating basic charts and visualisations;
- `seaborn` offers advanced statistical visualisations with improved aesthetics;
 - `pd.set_option('display.max_columns', None)` guarantees that all columns in the dataset are displayed without truncation for better data inspection and understanding.

 Reason for importing these libraries at the start of the notebook guarantees that all necessary tools are available for ensuing preprocessing, analysis, and model construction tasks and helps maintain an ordered workflow.

In [46]:
print("-" * 40)
print(" Import Required Libraries ")
print(" ***************************** ")
print("Libraries Imported Successfully!!!")
print("-" * 40)
import pandas as pd
import numpy as np

# Visualization Libraries
import matplotlib.pyplot as plt
import seaborn as sns

# Display Settings
pd.set_option('display.max_columns', None)

----------------------------------------
 Import Required Libraries 
 ***************************** 
Libraries Imported Successfully!!!
----------------------------------------


#Load Dataset

The combined air quality dataset, which includes meteorological variables and air pollution measurements gathered from several monitoring stations during a given time period, is loaded into the Python environment in this section for additional analysis and processing. The dataset includes meteorological variables and measurements of air pollutants that were gathered over a given time period from several monitoring stations.

The CSV file is read by the `pandas` library and stored in a DataFrame called `df`, which offers an organised tabular format for effective data handling and analysis.

Because it enables the data to be examined, cleansed, analysed, and visualised throughout the project, effectively loading the dataset is a crucial first step. Before beginning preprocessing, exploratory data analysis (EDA), and machine learning activities, this phase also guarantees that the dataset is properly connected to the notebook.

In [43]:
print("-" * 40)
print(" Load Dataset")
print(" ***************************** ")
print("Dataset Loaded Successfully!!!")
print("-" * 40)

df = pd.read_csv("combined_air_quality.csv")

----------------------------------------
 Load Dataset
 ***************************** 
Dataset Loaded Successfully!!!
----------------------------------------


#Dataset Shape

By counting the total number of rows and columns in the dataset, this part gives a rudimentary picture of its structure. Determining the quantity of the data and the number of variables available for analysis depends on an understanding of the dataset shape.

The dataset dimensions, where rows stand for observations and columns for characteristics or variables, are retrieved using the `.shape` function. To enhance presentation and readability, the findings are then shown in a structured table.

In [36]:
print(" ============================== ")
print("📊 Dataset Overview")
print(" ============================== ")

rows, columns = df.shape

shape_df = pd.DataFrame({
    "Description": ["Number of Rows", "Number of Columns"],
    "Count": [rows, columns]
})

shape_df.style \
    .set_properties(**{
        'background-color': '#f8f9fa',
        'color': '#000000',
        'border-color': 'black',
        'text-align': 'center'
    }) \
    .set_table_styles([
        {
            'selector': 'th',
            'props': [
                ('background-color', '#0d6efd'),
                ('color', 'white'),
                ('font-size', '14px'),
                ('text-align', 'center')
            ]
        }
    ])

📊 Dataset Overview


,Description,Count
0,Number of Rows,140256
1,Number of Columns,19


#Displaying First 5 Records

To get a general idea of the data structure, column names, and sample values, this section shows the dataset's first five rows. Previewing the dataset enables early detection of possible problems, such as missing numbers or incorrect formatting, and helps confirm that the data has been loaded correctly.

The first five rows are retrieved using the `head()` function, and table styling is employed to enhance the notebook's readability and presentation quality.

In [41]:
print(" ============================== ")
print(" Display First 5 Rows ")
print(" ============================== ")

df.head().style \
    .background_gradient(cmap='Blues') \
    .set_properties(**{
        'text-align': 'center',
        'border': '1px solid black'
    })

 Display First 5 Rows 


,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station,datetime
0,1,2013,3,1,0,9.000000,9.000000,3.000000,17.000000,300.000000,89.000000,-0.500000,1024.500000,-21.400000,0.000000,NNW,5.700000,Dongsi,2013-03-01 00:00:00
1,2,2013,3,1,1,4.000000,4.000000,3.000000,16.000000,300.000000,88.000000,-0.700000,1025.100000,-22.100000,0.000000,NW,3.900000,Dongsi,2013-03-01 01:00:00
2,3,2013,3,1,2,7.000000,7.000000,nan,17.000000,300.000000,60.000000,-1.200000,1025.300000,-24.600000,0.000000,NNW,5.300000,Dongsi,2013-03-01 02:00:00
3,4,2013,3,1,3,3.000000,3.000000,5.000000,18.000000,nan,nan,-1.400000,1026.200000,-25.500000,0.000000,N,4.900000,Dongsi,2013-03-01 03:00:00
4,5,2013,3,1,4,3.000000,3.000000,7.000000,nan,200.000000,84.000000,-1.900000,1027.100000,-24.500000,0.000000,NNW,3.200000,Dongsi,2013-03-01 04:00:00


#Coloumn Names

In order to comprehend the variables contained in the air quality data, this section shows all column names that are present in the dataset. In order to identify pollutant measurements, climatic characteristics, and time-related variables that will be utilised during data analysis and model construction, it is crucial to identify column names.

To enhance readability and give a more comprehensive picture of the dataset structure, the column names are arranged in a structured table.





In [80]:
print("-" * 40)
print("  Column Names ")
print("-" * 40)

columns_df = pd.DataFrame({
    "Column Names": df.columns
}).T

columns_df.style \
    .background_gradient(cmap='viridis') \
    .set_properties(**{
        'text-align': 'left',
        'font-size': '13px'
    })

----------------------------------------
  Column Names 
----------------------------------------


,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18
Column Names,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station,datetime


# Duplicates Values

In order to guarantee data integrity prior to additional analysis, this component finds and counts duplicate rows in the dataset. Because duplicate data introduce bias or over-represent some observations, they can have a negative impact on statistical results and machine learning models.

We can determine whether data cleaning procedures, like eliminating duplicates, are required by computing the total number of duplicate rows. This guarantees that every record in the dataset is distinct and appropriately contributes to the analysis procedure.

For improved readability and interpretation, the outcome is presented in an organised, aesthetically pleasing fashion.

In [94]:
duplicate_df = pd.DataFrame({
    "Metric": ["Total Duplicate Rows"],
    "Count": [df.duplicated().sum()]
})

duplicate_df.style \
    .background_gradient(cmap='Oranges') \
    .set_properties(**{
        'text-align': 'center',
        'border': '1px solid black'
    })

,Metric,Count
0,Total Duplicate Rows,0


## Datetime Structure

Datetime-related variables were examined to guarantee proper temporal formatting for time-series analysis, and duplicate observations were eliminated from the dataset to enhance data quality and avoid biased analytical conclusions.To guarantee correctness and avoid duplication bias in statistical results, duplicate records were eliminated.

To help you understand the structure of time-based data, this section presents a sample of the dataset's datetime-related columns. Correctly formatted temporal variables are crucial for trend analysis, seasonal pattern detection, and forecasting activities because this is a time-series dataset.

We can verify that the dataset is appropriately organised for time-based analysis by looking over the `datetime`, `year`, `month`, and `day` columns. This stage contributes to the accuracy and dependability of subsequent processes including grouping, resampling, and time-based visualisation.

To enhance readability and presentation, the data is also structured in a clear tabular view.

In [62]:
df[['datetime', 'year', 'month', 'day']].head(10).style \
    .set_properties(**{'text-align': 'center'}) \
    .set_table_styles([{
        'selector': 'th',
        'props': [('text-align', 'center')]
    }])

,datetime,year,month,day
0,2013-03-01 00:00:00,2013,3,1
1,2013-03-01 01:00:00,2013,3,1
2,2013-03-01 02:00:00,2013,3,1
3,2013-03-01 03:00:00,2013,3,1
4,2013-03-01 04:00:00,2013,3,1
5,2013-03-01 05:00:00,2013,3,1
6,2013-03-01 06:00:00,2013,3,1
7,2013-03-01 07:00:00,2013,3,1
8,2013-03-01 08:00:00,2013,3,1
9,2013-03-01 09:00:00,2013,3,1


# Missing Values

A sample of the dataset and a brief summary of data completeness are presented in this section. To evaluate the dataset's quality and spot any data gaps that can interfere with additional analysis or model construction, the total number of missing values is computed.

Understanding missing values is crucial because improper handling of partial data can affect statistical outcomes and lower model accuracy.

Additionally, the dataset's first five rows are shown to give a brief overview of the data's structure, column values, and general format. To make it simpler to visually examine the dataset within the notebook, a colour gradient is used to draw attention to missing or lower-value patterns.

In [77]:
print("-" * 40)
print("\nMissing Values:", df.isnull().sum().sum())
print("-" * 40)
display(df.head()
    .style
    .background_gradient(cmap='Reds', axis=1)
    .set_properties(**{
        'text-align': 'center',
        'border': '1px solid black'
    })
    .set_table_styles([{
        'selector': 'th',
        'props': [
            ('background-color', 'black'),
            ('color', 'white'),
            ('text-align', 'center')
        ]
    }]))

----------------------------------------

Missing Values: 601
----------------------------------------


,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station,datetime
0,1,2013,3,1,0,9.000000,9.000000,3.000000,17.000000,300.000000,89.000000,-0.500000,1024.500000,-21.400000,0.000000,NNW,5.700000,Dongsi,2013-03-01 00:00:00
1,2,2013,3,1,1,4.000000,4.000000,3.000000,16.000000,300.000000,88.000000,-0.700000,1025.100000,-22.100000,0.000000,NW,3.900000,Dongsi,2013-03-01 01:00:00
2,3,2013,3,1,2,7.000000,7.000000,15.811354,17.000000,300.000000,60.000000,-1.200000,1025.300000,-24.600000,0.000000,NNW,5.300000,Dongsi,2013-03-01 02:00:00
3,4,2013,3,1,3,3.000000,3.000000,5.000000,18.000000,1192.068962,57.692177,-1.400000,1026.200000,-25.500000,0.000000,N,4.900000,Dongsi,2013-03-01 03:00:00
4,5,2013,3,1,4,3.000000,3.000000,7.000000,47.129503,200.000000,84.000000,-1.900000,1027.100000,-24.500000,0.000000,NNW,3.200000,Dongsi,2013-03-01 04:00:00


# Data Types

In order to comprehend how the variables are kept and how they might be utilised for analysis, this section identifies the data types of each column in the dataset. Understanding data types is crucial since it guarantees that the right preprocessing methods are used and helps identify which columns are numerical, categorical, or datetime-based.

For instance, whereas categorical columns might need to be encoded before additional processing, numerical columns can be used for statistical analysis and machine learning models.

To enhance readability and facilitate the interpretation of the dataset structure, the results are presented in an organised fashion with visual styling.

In [81]:
print("-" * 40)
print("  Data Types ")
print("-" * 40)

dtype_df = pd.DataFrame(df.dtypes, columns=['Data Type']).T

dtype_df.style \
    .background_gradient(cmap='coolwarm') \
    .set_properties(**{
        'text-align': 'center',
        'border': '1px solid black'
    })

----------------------------------------
  Data Types 
----------------------------------------


,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station,datetime
Data Type,int64,int64,int64,int64,int64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,object,float64,object,object


# Data Type Count

The number of columns that belong to each type of data in the dataset is summarised in this section. This summary makes it easier to immediately comprehend the overall structure of the dataset in terms of numerical, categorical, and other variable kinds rather than focusing on specific column types.

This helps determine whether encoding, type conversion, or scaling may be necessary prior to doing analysis or developing machine learning models, which is helpful during the data preprocessing step. To improve readability and clarity, the output is also formatted as a structured table.

In [17]:
df.dtypes.value_counts().reset_index().rename(
    columns={"index": "Data Type", 0: "Count"}
)

,Data Type,count
0,float64,11
1,int64,5
2,object,3


# Verifying Missing Values

After the initial inspection or preprocessing stages, this phase is used to double-check and confirm whether the dataset contains any missing values. By determining whether any missing values are still present and may require additional treatment prior to analysis or model construction, it aids in ensuring data quality.

Because it guarantees that the dataset is complete and consistent and lowers the possibility of errors or bias in subsequent phases like statistical analysis or machine learning modelling, verifying missing values is a crucial step in the data cleaning process.

To facilitate the interpretation of the remaining missing data across all variables, the findings are presented in an organised fashion with a visual gradient.

In [92]:
print("-" * 40)
print("  Verify Missing Values ")
print("-" * 40)

verify_missing = pd.DataFrame(
    df.isnull().sum(),
    columns=['Remaining the Missing Values']
)

verify_missing.style \
    .background_gradient(cmap='Reds') \
    .set_properties(**{
        'text-align': 'center'
    })

----------------------------------------
  Verify Missing Values 
----------------------------------------


,Remaining the Missing Values
No,0
year,0
month,0
day,0
hour,0
PM2.5,0
PM10,0
SO2,0
NO2,0
CO,0


** Pollutant concentration variations, missing values, and possible correlations between pollution and meteorological variables are all indicated by preliminary study.

Significant variance in pollutant levels, especially PM2.5 and PM10, is shown in the statistical summary, reflecting varying air quality conditions throughout the dataset.

The table below provides a thorough statistical summary of the air quality dataset, which contains hourly observations from 2013 to 2017. It comprises climatic conditions, air pollution concentrations, and temporal variables. In order to comprehend environmental trends and get the dataset ready for additional analysis, each column offers information about the distribution, variability, and quality of the data.

The number of valid observations for each variable is displayed in the "count" row. A handful of pollutants, including PM2.5, PM10, SO2, NO2, CO, and O3, have missing data, even though the majority of variables have roughly 140,256 entries. This underlines the need for preprocessing, such as imputation or the elimination of missing values, and indicates data loss that is probably caused by sensor malfunction, environmental interference, or recording problems.

The hour variable verifies a full 24-hour cycle of observations, which is crucial for time-series analysis, while the mean year value (2014.66) shows that data is distributed throughout the entire time range. The temporal variables (year, month, day, and hour) confirm that the dataset spans multiple years (2013–2017) with uniform hourly recordings.

The mean concentration of PM2.5 (fine particulate matter) is 77.48, with values ranging from 2 to 882. The high maximum value and large standard deviation (78.36) show strong variability and extreme pollution episodes; the distribution is right-skewed because the median (50%) is lower than the mean, indicating that occasional severe pollution events have a significant impact on average levels.

Likewise, PM10 (coarse particulate matter) exhibits extremely significant pollution spikes, with a mean of 101.40 and a maximum of 999. Both typical air conditions and severe dust or industrial pollution occurrences are suggested by the large range. A skewed distribution is also supported by the discrepancy between the mean and median (79).

The gaseous pollutants exhibit varying degrees of variability: NO2 (nitrogen dioxide) has a mean of 47.13 and a maximum of 270, reflecting traffic-related emissions with moderate variability and a comparatively higher proportion of missing data compared to other pollutants; SO2 (sulphur dioxide) has a relatively low mean (15.81) but a high maximum (315), suggesting mostly stable conditions with sporadic industrial emission spikes.

O3 (Ozone) shows a mean of 57.69 but a very high maximum value (1071), which may indicate unusual atmospheric reactions or potential outliers/sensor anomalies. Ozone levels are usually influenced by sunlight and temperature, which explains their variability. CO (Carbon Monoxide) shows a very high mean value (1192) with a maximum reaching 10,000, indicating significant variability and potential pollution peaks caused by combustion activities like traffic congestion and fuel burning. The large standard deviation confirms that CO levels vary greatly over time.

Environmental context is provided by the meteorological variables. Temperature (TEMP) ranges from -19.9°C to 41.4°C with a mean of 13.35°C, indicating that all seasonal variations are captured in the dataset. Atmospheric pressure (PRES) stays relatively stable around 1009 hPa with minor fluctuations, which is typical for natural atmospheric conditions.

Rainfall (RAIN) has a very low mean (0.064), meaning that most hours have no rainfall but occasionally heavy rain events occur (up to 72.5), which can significantly affect air pollution levels by washing out particles. Dew Point (DEWP) has a mean of 2.32°C with a wide range, including very low values (-43.4°C), indicating dry atmospheric conditions during certain periods.

Lastly, Wind Speed (WSPM) has a mean of 1.77 m/s, which indicates usually low wind conditions. While greater wind speeds aid in dispersion and enhance air quality, lower wind speeds can lead to pollutant concentration.

Overall, the dataset demonstrates that while meteorological variables offer steady environmental background, air pollution variables are severely skewed with extreme values and missing records. These features emphasise how crucial it is to clean the data, handle outliers, and do additional correlation analysis before creating predictive models.


In [93]:
print("-" * 40)
print("  Statistical Summary ")
print("-" * 40)

df.describe().style \
    .background_gradient(cmap='PuBu') \
    .set_properties(**{
        'text-align': 'center',
        'font-size': '12px'
    })

----------------------------------------
  Statistical Summary 
----------------------------------------


,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,WSPM
count,140256.000000,140256.000000,140256.000000,140256.000000,140256.000000,140256.000000,140256.000000,140256.000000,140256.000000,140256.000000,140256.000000,140256.000000,140256.000000,140256.000000,140256.000000,140256.000000
mean,17532.500000,2014.662560,6.522930,15.729637,11.500000,77.481501,101.399508,15.811354,47.129503,1192.068962,57.692177,13.347028,1009.939314,2.328893,0.064937,1.768796
std,10122.141000,1.177201,3.448715,8.800123,6.922211,77.497957,89.001024,21.608348,32.420124,1068.882395,55.416629,11.500275,10.474478,13.848788,0.826344,1.252011
min,1.000000,2013.000000,1.000000,1.000000,0.000000,2.000000,2.000000,0.285600,1.026500,100.000000,0.214200,-19.900000,982.400000,-43.400000,0.000000,0.000000
25%,8766.750000,2014.000000,4.000000,8.000000,5.750000,20.000000,35.000000,2.000000,22.000000,500.000000,14.000000,3.000000,1001.500000,-9.200000,0.000000,0.900000
50%,17532.500000,2015.000000,7.000000,16.000000,11.500000,55.000000,81.000000,7.000000,41.000000,900.000000,47.000000,14.300000,1009.550000,2.800000,0.000000,1.400000
75%,26298.250000,2016.000000,10.000000,23.000000,17.250000,106.000000,139.000000,19.000000,64.000000,1400.000000,80.000000,23.100000,1018.100000,15.000000,0.000000,2.200000
max,35064.000000,2017.000000,12.000000,31.000000,23.000000,882.000000,999.000000,315.000000,270.000000,10000.000000,1071.000000,41.400000,1042.000000,29.100000,72.500000,12.900000
